ArmPkg/Drivers/SmmuDxe/SmmuDxe.c

```diff
@@ -0,0 +1,774 @@
+}
+
+/**
+  Configure the SMMUv3 based on the provided configuration per the SmmuV3 specification.
+
+  @param [in] SmmuInfo    Pointer to the SMMU_INFO structure.
+  @param [in] SmmuConfig  Pointer to the SMMU configuration.
+
+  @retval EFI_SUCCESS     Success.
+  @retval Others          Failure.
+**/
+EFI_STATUS
+EFIAPI
+SmmuV3Configure (
+  IN SMMU_INFO    *SmmuInfo,
+  IN SMMU_CONFIG  *SmmuConfig
+  )
+{
+  EFI_STATUS                 Status;
+  UINT32                     STLog2Size;
+  UINT32                     STSize;
+  UINT32                     CommandQueueLog2Size;
+  UINT32                     EventQueueLog2Size;
+  UINT8                      ReadWriteAllocationHint;
+  SMMUV3_STRTAB_BASE         StrTabBase;
+  SMMUV3_STRTAB_BASE_CFG     StrTabBaseCfg;
+  SMMUV3_STREAM_TABLE_ENTRY  *StreamTablePtr;
+  SMMUV3_CMDQ_BASE           CommandQueueBase;
+  SMMUV3_EVENTQ_BASE         EventQueueBase;
+  SMMUV3_STREAM_TABLE_ENTRY  TemplateStreamEntry;
+  SMMUV3_CR0                 Cr0;
+  SMMUV3_CR1                 Cr1;
+  SMMUV3_CR2                 Cr2;
+  SMMUV3_IDR0                Idr0;
+  SMMUV3_GERROR              GError;
+  SMMUV3_CMD_GENERIC         Command;
+  VOID                       *CommandQueue;
+  VOID                       *EventQueue;
+
+  if ((SmmuConfig->Config.SmmuNode.SmmuNode.Flags & EFI_ACPI_IORT_SMMUv3_FLAG_COHAC_OVERRIDE) != 0) {
+    ReadWriteAllocationHint = 0x1;
+  } else {
+    ReadWriteAllocationHint = 0x0;
+  }
+
+  GError.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase, SMMU_GERROR);
+  ASSERT (GError.AsUINT32 == 0);
+
+  // Disable SMMU before configuring
+  Status = SmmuV3DisableTranslation (SmmuInfo->SmmuBase);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Error SmmuV3Disable: SmmuBase=0x%lx\n", SmmuInfo->SmmuBase));
+    goto End;
+  }
+
+  Status = SmmuV3DisableInterrupts (SmmuInfo->SmmuBase, TRUE);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Error SmmuV3DisableInterrupts: SmmuBase=0x%lx\n", SmmuInfo->SmmuBase));
+    goto End;
+  }
+
+  // Only Index 16 is being used AFAIK
+  StreamTablePtr                = SmmuV3AllocateStreamTable (SmmuInfo, SmmuConfig, &STLog2Size, &STSize);
+  SmmuInfo->StreamTable         = StreamTablePtr;
+  SmmuInfo->StreamTableSize     = STSize;
+  SmmuInfo->StreamTableLog2Size = STLog2Size;
+  ASSERT (StreamTablePtr != NULL);
+
+  SmmuInfo->PageTableRoot = PageTableInit ();
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Error PageTableInit: SmmuBase=0x%lx\n", SmmuInfo->SmmuBase));
+    goto End;
+  }
+
+  // Build default STE template
+  Status = SmmuV3BuildStreamTable (SmmuInfo, SmmuConfig, &TemplateStreamEntry);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Error SmmuV3BuildStreamTable: SmmuBase=0x%lx\n", SmmuInfo->SmmuBase));
+    goto End;
+  }
+
+  // Load default STE values
+  // Only Index 16 is being used AFAIK
+  for (UINT32 i = 0; i < SMMUV3_COUNT_FROM_LOG2 (STLog2Size); i++) {

```

Comment:
`i` should be defined above with other variables and called `Index`, by style guidelines

Reply:
fixed